Tribler/Policies/BoostingSource.py

```diff
@@ -0,0 +1,639 @@
+                torrent_id = self.torrent_db.getTorrentID(torrent.infohash)
+                if torrent_id:
+                    torrent.update_torrent_id(torrent_id)
+
+            if not self.session.has_collected_torrent(torrent.infohash):
+                files = []
+                trackers = []
+
+                # see if we have most info in our tables
+                if isinstance(torrent, RemoteTorrent):
+                    torrent_id = self.torrent_db.getTorrentID(torrent.infohash)
+                else:
+                    torrent_id = torrent.torrent_id
+
+                trackers.extend(self.torrent_db.getTrackerListByTorrentID(torrent_id))
+
+                if 'DHT' in trackers:
+                    trackers.remove('DHT')
+                if 'no-DHT' in trackers:
+                    trackers.remove('no-DHT')
+
+                # replacement # self.downloadTorrentfileFromPeers(torrent, None)
+                if self.session.has_download(torrent.infohash):
+                    return False
+
+                if torrent.query_candidates is None or len(torrent.query_candidates) == 0:
+                    self.session.download_torrentfile(torrent.infohash, None, 0)
+                else:
+                    for candidate in torrent.query_candidates:
+                        self.session.download_torrentfile_from_peer(candidate, torrent.infohash, None, 0)
+
+                torrent = NotCollectedTorrent(torrent, files, trackers)
+
+            else:
+                tdef = TorrentDef.load_from_memory(self.session.get_collected_torrent(torrent.infohash))
+
+                if torrent.torrent_id <= 0:
+                    del torrent.torrent_id
+
+                torrent = CollectedTorrent(torrent, tdef)
+
+        # replacement # self.library_manager.addDownloadState(torrent)
+        for dl_state in self.dslist:
+            torrent.addDs(dl_state)
+
+        # return
+        if callback is not None:
+            callback(torrent)
+        else:
+            return torrent
+
+    @staticmethod
+    def get_instance(*args, **kw):
+        """
+        get single instance of TorrentManagerCM
+        """
+        if TorrentManagerCM.__single is None:
+            TorrentManagerCM(*args, **kw)
+        return TorrentManagerCM.__single
+
+    @staticmethod
+    def del_instance():
+        """
+        resetting, then deleting single instance
+        """
+        TorrentManagerCM.__single = None
+
+
+class BoostingSource(object):
+    """
+    Base class for boosting source. For now, it can be RSS, directory, and channel
+    """
+
+    def __init__(self, session, source, boost_settings, callback):
+        self.session = session
+        self.channelcast_db = session.lm.channelcast_db
+
+        self.torrents = {}
+        self.source = source
+        self.interval = boost_settings.source_interval
+        self.max_torrents = boost_settings.max_torrents_per_source
+        self.callback = callback
```

Comment:
It called `on_torrent_insert` in `BoostingManager`. To put a ready torrent into the manager.